### PR TITLE
Device-controller

### LINKED ISSUE #77: Create DeviceController
- [x] Create CastDeviceVoteRequest

- [x] Create deviceController, takes in number of yes and number of no 

- [x] Create method in voteService that deviceController can use and takes in a number of yes and number of no, creates these votes and persists them. 

- [x] Let Device extend Voter

### LINKED ISSUE #81: votes do not get added to poll when saved in voteRepository
`vote.setPoll(poll); voteRepository.save(vote);` does not link the vote to given poll: `pollRepository.getPollById(poll.getId()).get().getVotes().size()` fails and the poll have zero votes. 

**Solution**    
We need to add functionality to add votes to poll and then persist it. The same for user. (I am on it) 😎

### PR DESCRIPTION
**Pull request by JennStro**
Closes #81  and #77 

Sorry for lang PR, men her måtte det fiskes litt underveis. 😬 

**Changed**
* Moved GuestController to own package for IT tests 

**Added**
* A poll can now get more votes after persisted in db  by `poll.addVote(vote); pollRepository.save(poll)` (save updates the poll if it already excists)
* A user can now add polls after persisted in db by `user.addPoll(poll); userRepository.save(user)`
* Method to delete a poll in user due to  [this answer](https://stackoverflow.com/questions/29172313/spring-data-repository-does-not-delete-manytoone-entity) . 
* Fixes #81 Vote does not get linked to poll when persisted: Setters in poll and vote so that vote.setPoll(poll) and poll.addVote(vote) both can be used due to [this explanation](https://stackoverflow.com/questions/13370221/persistentobjectexception-detached-entity-passed-to-persist-thrown-by-jpa-and-h)   Also did this for Voter and Vote. **We need to fix this for Poll and User as well**.
* Tests for endpoint. Will make IT in another PR.

### CODE REVIEW

File: src/main/java/no/hvl/dat250/h2020/group5/service/UserService.java [hunk 1]
```diff
@@ -17,81 +17,77 @@
...
+      return false;
     }
 
-    public UserResponse getUser(Long userId) {
-        Optional<User> user = userRepository.findById(userId);
-        if (user.isEmpty()) {
-            return null;
-        }
-        return new UserResponse(user.get());
+    List<Vote> votes = user.get().getVotes();
```
Comment (oddhus) [inline src/main/java/no/hvl/dat250/h2020/group5/service/UserService.java:42]: Lurt! 👍

File: src/main/java/no/hvl/dat250/h2020/group5/service/VoteService.java [hunk 2]
```diff
@@ -1,90 +1,113 @@
...
+    Optional<Poll> poll = pollRepository.findById(voteRequestFromDevice.getPollId());
+
+    if (poll.isPresent()) {
+      List<Vote> votes = voteRequestFromDevice.getVotes();
+      voteRepository.saveAll(votes);
+      return voteRepository.saveAll(
+          votes.stream().peek(vote -> vote.setPoll(poll.get())).collect(Collectors.toList()));
     }
+    return null;
+  }
```
Comment (oddhus) [inline src/main/java/no/hvl/dat250/h2020/group5/service/VoteService.java:112]: Flott løsning! 😄

File: src/main/lombok/no/hvl/dat250/h2020/group5/entities/Poll.java [hunk 3]
```diff
@@ -17,47 +17,64 @@
...
+   * @return True if vote is added to this poll
+   */
+  public boolean addVote(Vote vote) {
+    if (votes.contains(vote) || vote.getPoll() != null) {
+      return false;
     }
+    this.votes.add(vote);
+    vote.setPoll(this);
+    return true;
+  }
```
Comment (oddhus) [inline src/main/lombok/no/hvl/dat250/h2020/group5/entities/Poll.java]: Fint at vi fikk denne metoden! 🥇 Mindre sjanse for å gjøre feil

File: src/main/lombok/no/hvl/dat250/h2020/group5/entities/User.java [hunk 4]
```diff
@@ -15,32 +15,41 @@
...
+  }
+
+  public void addPoll(Poll poll) {
+    poll.setPollOwner(this);
+    this.userPolls.add(poll);
+  }
+
+  public boolean deletePoll(Poll poll) {
+    return userPolls.remove(poll);
+  }
```
Comment (oddhus) [inline src/main/lombok/no/hvl/dat250/h2020/group5/entities/User.java:54]: Godt å ha disse på plass også!

File: src/main/lombok/no/hvl/dat250/h2020/group5/entities/Vote.java [hunk 5]
```diff
@@ -11,29 +11,61 @@
...
+      this.voter = null;
+      return true;
+    }
+    if (voter.getVotes().contains(this)) {
+      return false;
     }
+    voter.addVote(this);
+    this.voter = voter;
+    return true;
+  }
```
Comment (oddhus) [inline src/main/lombok/no/hvl/dat250/h2020/group5/entities/Vote.java:70]: 💯

File: src/test/java/no/hvl/dat250/h2020/group5/controllers/VotingDeviceControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    this.noVote = new Vote().answer(AnswerType.NO);
+    this.poll = new Poll();
+
+    yesVote.setPoll(poll);
+    noVote.setPoll(poll);
+  }
+
+  @Test
+  public void shouldSaveOneYesVoteTest() throws Exception {
+    when(voteService.saveVotesFromDevice(any())).thenReturn(Collections.singletonList(yesVote));
```
Comment (oddhus) [inline src/test/java/no/hvl/dat250/h2020/group5/controllers/VotingDeviceControllerTest.java:46]: Går eventuelt an å ha any(VoteRequestFromDevice.class) her, men har nok ikke mye å si. Gode tester 🚀 👍

File: src/main/java/no/hvl/dat250/h2020/group5/requests/VoteRequestFromDevice.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    List<Vote> votes = new ArrayList<>();
+    for (int i = 0; i < numberOfYes; i++) {
+      votes.add(new Vote().answer(AnswerType.YES));
+    }
+    for (int i = 0; i < numberOfNo; i++) {
+      votes.add(new Vote().answer(AnswerType.NO));
+    }
+    return votes;
+  }
+}
```
Comment (oddhus) [inline src/main/java/no/hvl/dat250/h2020/group5/requests/VoteRequestFromDevice.java:37]: Er det slik at du ikke setter pollId her fordi de må lagres først? 🕵️

**Review by oddhus — DISMISSED** [review]
Det ser veldig bra ut! Du nøstet opp i flere problemer og fikset dem! 🚀 🔥 😄

File: src/main/java/no/hvl/dat250/h2020/group5/requests/VoteRequestFromDevice.java (hunk 7 above)
Comment (JennStro) [reply src/main/java/no/hvl/dat250/h2020/group5/requests/VoteRequestFromDevice.java:37]: Ja, for hvis vote.setPoll(poll) kalles på her, så kaller den videre på poll.addVote(vote), men vote har ikke fått noen ID enda siden de ikke er lagret, og da blir vote_id lik null. :)

**Comment by JennStro** [comment]
> **We need to fix this for Poll and User as well**.

Lager et issue på dette.

**Review by oddhus — APPROVED** [review]
Supert 💯

**Pull request merged**